### PR TITLE
scripts: add script for batch-killing-off pgs's past_intervals

### PR DESCRIPTION
**Pull request by Yan-waller**
Signed-off-by: Yan Jun <yan.jun8@zte.com.cn>

### CODE REVIEW

File: scripts/remove_past_intervals.sh [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+OSD_NUM=$1
+OSD_PATH=/var/lib/ceph/osd/ceph-$OSD_NUM
+shift
+
+test -e $OSD_PATH || echo "osd.$OSD_NUM not exist."
+
+TOOL=`which ceph-objectstore-tool`
+
+echo "Remove all the pg past intervals of osd.$OSD_NUM"
```
Comment (xiexingguo) [inline scripts/remove_past_intervals.sh]: start to remove all past_intervals for osd.$OSD_NUM

File: scripts/remove_past_intervals.sh [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+PGS=`$TOOL --data-path $OSD_PATH --skip-journal-replay --op list-pgs`
+for pg in $PGS
+do
+  echo "osd.$OSD_NUM $pg past intervals removing..."
```
Comment (xiexingguo) [inline scripts/remove_past_intervals.sh]: removing $pg 's past_intervals

File: scripts/remove_past_intervals.sh [hunk 3]
```diff
@@ -0,0 +1,31 @@
+#!/bin/bash
+# Program:
+# 		  User can remove pg's past intervals for a specified OSD
```
Comment (xiexingguo) [inline scripts/remove_past_intervals.sh]: Remove all pgs' past_intervals for a specified OSD.
Please use with caution as there is no turning back.

**Review by xiexingguo — COMMENTED** [review]
Excellent!

**Review by xiexingguo — APPROVED** [review]

File: scripts/remove_past_intervals.sh [hunk 4]
```diff
@@ -0,0 +1,35 @@
+#!/bin/bash
+# Program:
+#       Remove all pgs' past_intervals for a specified OSD, please use with
+#       caution as there is no turning back.
+# History:
+# 2018/10/13		yanjun			create
+
+case $1 in
+  [0-9]*)
+        ;;
```
Comment (runsisi) [inline scripts/remove_past_intervals.sh]: indent with 2 spaces

```bash
case $1 in
  *[!0-9]* | '')
    echo -e "Usage:\n\t $0 <osdid>\t numerical value, such as 0,1,2..."
    exit 1
    ;;
  *)
    ;;
esac
```

may be better, we need to error out when called with wrong argument, e.g., `13a`

File: scripts/remove_past_intervals.sh [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+if [ ! -e $OSD_PATH ]; then
+  echo "Error: osd.$OSD_NUM not exist."
+  exit 1
+fi
+
+TOOL=`which ceph-objectstore-tool`
+
+echo "start to remove all past_intervals for osd.$OSD_NUM"
+PGS=`$TOOL --data-path $OSD_PATH --skip-journal-replay --op list-pgs`
```
Comment (runsisi) [inline scripts/remove_past_intervals.sh:40]: check return value before continue

File: scripts/remove_past_intervals.sh [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+for pg in $PGS
```
Comment (runsisi) [inline scripts/remove_past_intervals.sh]:
```bash
for .. in ..;  do
  ...
done
```
keep the same style as `if .. fi`, may be better

**Comment by runsisi** [comment]
add a dry-run option should be better

generally lgtm

File: scripts/remove_past_intervals.sh [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+        exit 1
+        ;;
+esac
+
+OSD_NUM=$1
+OSD_PATH=/var/lib/ceph/osd/ceph-$OSD_NUM
+shift
+
+if [ ! -e $OSD_PATH ]; then
+  echo "Error: osd.$OSD_NUM not exist."
```
Comment (xiexingguo) [inline scripts/remove_past_intervals.sh]: does not exist

**Comment by Yan-waller** [comment]
@xiexingguo @runsisi thanks for your comments, add an `usage` function and simply `--dry-run` option
```
Usage:
         remove_past_intervals.sh [OPTION] <osdid>       numerical osd id, such as 0,1,2...
Options
  -h, --help     show this help
  -n, --dry-run  perform a trial run with no changes made
```

File: scripts/remove_past_intervals.sh [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+  echo -e "Options\n  -h, --help\t show this help"
+  echo -e "  -n, --dry-run\t perform a trial run with no changes made"
+  exit $1
+}
+
+test $# -eq 0 && usage 1
+for opt in $@; do
+  case $opt in
+    "-n" | "--dry-run")
+      set -v
```
Comment (runsisi) [inline scripts/remove_past_intervals.sh]: i can not understand this :(

Comment (Yan-waller) [reply scripts/remove_past_intervals.sh]: I just simply print the script command lines (-v, --verbose)and do not execute them (-n, --noexec) here, I'm not very sure whether we should do something more.

**Comment by Yan-waller** [comment]
@runsisi passed `--dry-run` option to `ceph-objectstore-tool` which will handle it.

File: scripts/remove_past_intervals.sh [hunk 9]
```diff
@@ -0,0 +1,50 @@
+#!/bin/bash
+# Program:
+#       Remove all pgs' past_intervals for a specified OSD, please use with
+#       caution as there is no turning back.
+# History:
+# 2018/10/13		yanjun			create
```
Comment (runsisi) [inline scripts/remove_past_intervals.sh]: no need to add history in files, git history should be enough

Comment (xiexingguo) [reply scripts/remove_past_intervals.sh]: Good point.

Comment (Yan-waller) [reply scripts/remove_past_intervals.sh]: deleted it

**Review by runsisi — APPROVED** [review]

**Pull request merged**